frappe/desk/reportview.py

```diff
@@ -18,20 +19,157 @@
 @frappe.read_only()
 def get():
 	args = get_form_params()
+	return compress(execute(**args), args=args)
 
-	data = compress(execute(**args), args = args)
+@frappe.whitelist()
+@frappe.read_only()
+def get_list():
+	# uncompressed (refactored from frappe.model.db_query.get_list)
+	return execute(**get_form_params())
 
-	return data
+@frappe.whitelist()
+@frappe.read_only()
+def get_count():
+	args = get_form_params()
+	args.fields = ['{distinct}count(name) as total_count'.format(distinct = 'distinct ' if args.distinct=='true' else '')]
+	return execute(**args)[0].get('total_count')
 
 def execute(doctype, *args, **kwargs):
 	return DatabaseQuery(doctype).execute(*args, **kwargs)
 
 def get_form_params():
 	"""Stringify GET request parameters."""
 	data = frappe._dict(frappe.local.form_dict)
+	clean_params(data)
+	validate_args(data)
+	return data
+
+def validate_args(data):
+	parse_json(data)
+	setup_group_by(data)
+
+	validate_fields(data)
+	if data.filters:
+		validate_filters(data, data.filters)
+	if data.or_filters:
+		validate_filters(data, data.or_filters)
+
+	data.strict = None
 
-	is_report = data.get('view') == 'Report'
+	return data
+
+def validate_fields(data):
```

Comment:
Looks like all validators changing the data inline without returning the data. 

I prefer to go with a class here as we are trying to keep the state of the `data`. May be something like

q = QueryBuilder(data)
q.is_valid_query() - says the query is valid or not
q.data - returns me the constructed data that I can use to query

Reply:
`validate_fields` is also called independently from `frappe.client.get_list`